### PR TITLE
Refactoring leaves and calendar controllers

### PR DESCRIPTION
**Pull request by HarshwardhanSingh**

### CODE REVIEW

File: app/controllers/leaves_controller.rb [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    response = @client.remove_leave(params[:calendar_id], params[:id])
+    redirect_to :back
+  end
+
+  private
+    def auth_check
+      if !is_logged_in
+        redirect_to login_path
+      end
+    end
```
Comment (emilsoman) [inline app/controllers/leaves_controller.rb]: Should we move this to application controller ?

File: app/controllers/leaves_controller.rb [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    response = @client.edit_leave(params[:calendar_id],params[:id])
+    render locals:{ leave: JSON.parse(response.body) }
+  end
+
+  # Update an leave
+  def update
+    leave = {
+      'summary' => params[:summary],
+      'start' => {'dateTime' => Time.now.iso8601},
+      'end' => {'dateTime' => Time.now.iso8601}
```
Comment (emilsoman) [inline app/controllers/leaves_controller.rb]: When converting time to standardized iso8601 format, always convert the time in UTC:

`Time.now.utc.iso8601`

Comment (emilsoman) [inline app/controllers/leaves_controller.rb]: Why are both start and end time same ?

File: app/services/google_calendar_service.rb [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    @client.authorization.client_secret = ENV["GOOGLE_SECRET"]
+    if @client.authorization.refresh_token && @client.authorization.expired?
+      @client.authorization.fetch_access_token!
+    end
+    return @client
+  end
+
+  def list_calendars
+    service = @client.discovered_api('calendar' , 'v3')
+    response = @client.execute(api_method: service.calendar_list.list,
```
Comment (emilsoman) [inline app/services/google_calendar_service.rb]: Why do we need "response" variable here ? (Same for all methods below)

File: app/services/google_calendar_service.rb [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    @client.authorization.client_id = ENV["GOOGLE_CLIENT_ID"]
+    @client.authorization.client_secret = ENV["GOOGLE_SECRET"]
+    if @client.authorization.refresh_token && @client.authorization.expired?
+      @client.authorization.fetch_access_token!
+    end
+    return @client
+  end
+
+  def list_calendars
+    service = @client.discovered_api('calendar' , 'v3')
```
Comment (emilsoman) [inline app/services/google_calendar_service.rb]: I think we can move this logic to a private method named "service" to DRY this code.

File: app/services/google_calendar_service.rb [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+  def edit_calendar(calendar_id)
+    service = @client.discovered_api('calendar' , 'v3')
+    calendarId = Base64.urlsafe_decode64(calendar_id)
+    reponse = @client.execute(:api_method => service.calendars.patch,
+                  parameters: {'calendarId' => calendarId})
+  end
+
+  def update_calendar
+  end
```
Comment (emilsoman) [inline app/services/google_calendar_service.rb]: Can we remove this unused method ?

File: app/services/google_calendar_service.rb [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+                    'singleEvents': true,
+                    'maxResults': 10,
+                    'orderBy': 'startTime'},
+                  headers: { 'Content-Type' => 'application/json' })
+  end
+
+  def edit_calendar(calendar_id)
+    service = @client.discovered_api('calendar' , 'v3')
+    calendarId = Base64.urlsafe_decode64(calendar_id)
+    reponse = @client.execute(:api_method => service.calendars.patch,
```
Comment (emilsoman) [inline app/services/google_calendar_service.rb]: When using symbols as hash keys use new `{key: value}` syntax instead of old `{:key => value}` syntax.

File: app/views/calendars/edit.html.erb [hunk 7]
```diff
@@ -1,9 +1,9 @@
 <div class="container">
   <div class="col-md-6 col-md-offset-3">
     <div class="new-calendar-form">
-      <%= form_for(calendars_path(id: params[:id]), method: 'patch', url: calendars_path(id: params[:id])) do %>
+      <%= form_for(calendar_path(id: params[:id]), method: 'put', url: calendar_path(id: params[:id])) do %>
         <div class="form-group">
-          <%= text_field_tag("summary", nil, class: 'form-control', placeholder: 'Name of calendar') %>
+          <%= text_field_tag("summary", @calendar["summary"], class: 'form-control', placeholder: 'Name of calendar') %>
```
Comment (emilsoman) [inline app/views/calendars/edit.html.erb]: Where is instance variable `@calendar` coming from? Please refrain from using instance variables inside views, use locals instead.

File: app/views/calendars/index.html.erb [hunk 8]
```diff
@@ -2,7 +2,7 @@
   <div class="col-md-6 col-md-offset-3">
     <h3>Calendar List</h3>
     <hr>
-    <% @response.data["items"].each do |item| %>
+    <% response.data["items"].each do |item| %>
       <div class="calendar-card" style="background-color: <%= item['backgroundColor'] %>">
         <% id = Base64.urlsafe_encode64(item["id"]) %>
         <%= link_to item["summary"], calendar_path(id: id) %>
```
Comment (emilsoman) [inline app/views/calendars/index.html.erb]: Is there going to be a different PR for using a presenter to make the views logicless ?

File: spec/controllers/leaves_controller_spec.rb [hunk 9]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe LeavesController, :type => :controller do
+
```
Comment (emilsoman) [inline spec/controllers/leaves_controller_spec.rb]: Please add tests

File: spec/helpers/leaves_helper_spec.rb [hunk 10]
```diff
@@ -1,15 +1,15 @@
...
 #   describe "string concat" do
 #     it "concats two strings with spaces" do
 #       expect(helper.concat_strings("this","that")).to eq("this that")
 #     end
 #   end
 # end
-RSpec.describe EventsHelper, :type => :helper do
+RSpec.describe LeavesHelper, :type => :helper do
   pending "add some examples to (or delete) #{__FILE__}"
 end
```
Comment (emilsoman) [inline spec/helpers/leaves_helper_spec.rb]: Please remove this file and the related helper file if we're not using them.

File: app/views/calendars/index.html.erb (hunk 8 above)
Comment (HarshwardhanSingh) [inline app/views/calendars/index.html.erb]: @emilsoman yes, will do it today.

File: app/services/google_calendar_service.rb (hunk 4 above)
Comment (HarshwardhanSingh) [inline app/services/google_calendar_service.rb]: @emilsoman then we would have to call GoogleCalendarService.list_calendars instead of @client.list_calendars.
Also while naming it service and calling before_action to set @client it doesn't work.

Comment (emilsoman) [inline app/services/google_calendar_service.rb]: What I meant was a private method to DRY up line 17. Anyways, I think this shouldn't be in the controller at all. Which version of google calendar API we're using is the responsibility of the API wrapper (which encapsulates the implementation details of google calendar api request and response)

Comment (emilsoman) [inline app/services/google_calendar_service.rb]: @HarshwardhanSingh what I meant was something like this:

```ruby
private

def service
  @client.discovered_api('calendar', 'v3')
end
```

Comment (emilsoman) [inline app/services/google_calendar_service.rb]: `before_action` works only in Rails controllers - it's a Rails thing. Can't use that in pure ruby classes.

File: app/controllers/leaves_controller.rb (hunk 2 above)
Comment (HarshwardhanSingh) [inline app/controllers/leaves_controller.rb]: Fixed for create leave action. Will fix in next commit for update leave cz currently working on default datetime in views.

File: app/controllers/application_controller.rb [hunk 11]
```diff
@@ -8,7 +8,12 @@ def current_user
     @current_user ||= User.find(session[:user_id]) if session[:user_id]
   end
 
-  def is_logged_in
-    current_user || false
+  def is_logged_in?
+    if session[:user_id]
```
Comment (deepak) [inline app/controllers/application_controller.rb]: is an anti-pattern. 

we are just checking for a <variable>. just returning the variable is sufficient. 
if you are not comfortable returning the variable. and want to return a boolean for:

1. a better api
2. documentation

then can do a `double negation` ie. `!!session[:user_id]`

File: app/controllers/application_controller.rb [hunk 12]
```diff
@@ -8,7 +8,12 @@ def current_user
...
+      true
+    else
+      # redirect_to login_path
```
Comment (deepak) [inline app/controllers/application_controller.rb]: this is not the place for this comment.
this comment does not make sense, what if i want to log something if user is not logged in 

this is the problem with comments, they diverge from the actual code

File: app/controllers/calendars_controller.rb [hunk 13]
```diff
@@ -1,56 +1,58 @@
...
+    render locals:{calendars: response.data["items"]}
   end
 
   # Show calendar events
   def show
-    @response = @client.show_calendar(params[:id])
-    @items = @response.data['items']
+    response = @client.show_calendar(params[:id])
+    leaves = []
+    response.data['items'].each do |item|
```
Comment (deepak) [inline app/controllers/calendars_controller.rb]: can use map rather than each here

File: app/controllers/leaves_controller.rb [hunk 14]
```diff
@@ -0,0 +1,51 @@
+class LeavesController < ApplicationController
+  before_action :is_logged_in?, :connect
```
Comment (deepak) [inline app/controllers/leaves_controller.rb]: can be extracted into a base controller

File: app/controllers/leaves_controller.rb [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+
+  # Fetch all the leaves of current month of the logged in user
```
Comment (deepak) [inline app/controllers/leaves_controller.rb]: ratio of 1:1 for comment and code is not good
refactor method names so that comment is not required

File: app/controllers/leaves_controller.rb [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+
+  # Display new leave form
+  def new
+  end
+
+  # Create new leave
+  def create
+    leave = {
+      'summary' => params[:summary],
+      'start' => {'dateTime' => params[:start].to_time.utc.iso8601},
```
Comment (deepak) [inline app/controllers/leaves_controller.rb]: no strong preference. but in ruby, generally key names are like `date_time` not `dateTime`
this is because it is used as-is for a JSON response ?

File: app/controllers/leaves_controller.rb [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+      'end' => {'dateTime' => params[:end].to_time.utc.iso8601}
+    }
+    response = @client.create_leave(params[:calendar_id], leave)
+    redirect_to calendars_path
+  end
+
+  # Display Edit leave form
```
Comment (deepak) [inline app/controllers/leaves_controller.rb]: these comments are not required. the `edit` action displays the edit form is common knowledge for a rails app

File: app/presenters/calendar_presenter.rb [hunk 18]
```diff
@@ -0,0 +1,18 @@
...
+
+  def initialize(response)
+    @response = response
+  end
+
+  def name
+    JSON.parse(@response.body)["summary"]
+  end
+
+  def encoded_id
```
Comment (deepak) [inline app/presenters/calendar_presenter.rb]: i would call it `calendar_id` as helpers can be required on top of each other. they are all one big bag

File: app/controllers/leaves_controller.rb (hunk 16 above)
Comment (deepak) [inline app/controllers/leaves_controller.rb]: can use `params[:start].to_time(:utc)`

btw `to_time` is a weird little function
although does not affect us here - i think

```
date = "2012-12-13"
date.to_time(:utc)  #=> 2012-12-13 00:00:00 UTC
date.to_time #=> 2012-12-13 00:00:00 +0530
date.to_time.utc #=> 2012-12-12 18:30:00 UTC  # WTF
```

is a good idea to validate user inputs
but can be done later

File: app/presenters/leave_presenter.rb [hunk 19]
```diff
@@ -0,0 +1,21 @@
...
+  def id
+    @leave.id
+  end
+
+  def summary
+    @leave.summary
+  end
+
+  def start_date
+    @leave.start["dateTime"].to_time.strftime("%d-%B-%Y")
```
Comment (deepak) [inline app/presenters/leave_presenter.rb]: should this time be utc as well ? as in https://github.com/code-mancers/leave-tracker/pull/17/files#diff-7f181fd260b0c230c586b9d9684dcba7R12

also because you are interested only in the date, you can use `to_date` rather than  `to_time`

File: app/presenters/leave_presenter.rb [hunk 20]
```diff
@@ -0,0 +1,21 @@
...
+  end
+
+  def end_date
+    @leave.end["dateTime"].to_time.strftime("%d-%B-%Y")
```
Comment (deepak) [inline app/presenters/leave_presenter.rb]: this is a class rather than a module. which is good. 
which mean you are structure your code better.
this date extraction and formatting can be extracted into a private method

File: app/services/google_calendar_service.rb [hunk 21]
```diff
@@ -0,0 +1,126 @@
...
+  def initialize(oauth_token, refresh_token)
+    @client = Google::APIClient.new(:auto_refresh_token => true)
+    @client.authorization.access_token = oauth_token
+    @client.authorization.refresh_token = refresh_token
+    @client.authorization.client_id = ENV["GOOGLE_CLIENT_ID"]
+    @client.authorization.client_secret = ENV["GOOGLE_SECRET"]
+    if @client.authorization.refresh_token && @client.authorization.expired?
+      @client.authorization.fetch_access_token!
+    end
+    return @client
```
Comment (deepak) [inline app/services/google_calendar_service.rb]: two things why this explicit return is wrong, and should be removed:

1. it is redundant here as any return from a `initialize` method is ignored and is not returned
2. if this was not a `initialize` method then an explicit return at the end is bad form. the last expression is automatically returned. ie. just writing `@client` is sufficient

File: app/services/google_calendar_service.rb [hunk 22]
```diff
@@ -0,0 +1,126 @@
...
+  end
+
+  def list_calendars
+    @client.execute(api_method: service.calendar_list.list,
```
Comment (deepak) [inline app/services/google_calendar_service.rb]: `@client` can be a private attr_reader. ie. `def client; @client; end`
can be extracted out into a helper as method name and parameters are different

File: config/routes.rb [hunk 23]
```diff
@@ -16,7 +16,7 @@
   get '/dashboard', to: 'pages#dashboard', as: 'dashboard'
 
   resources :calendars do
-    resources :events
+    resources :leaves, as: 'leave'
```
Comment (deepak) [inline config/routes.rb]: this is not a singular resource, so would suggest to leave it as is
not worth fighting rails for personal style preferences

checkout singular resources. http://guides.rubyonrails.org/routing.html#singular-resources
if that is what you meant

File: app/controllers/api_controller.rb [hunk 24]
```diff
@@ -0,0 +1,12 @@
+class ApiController < ApplicationController
+  before_action :force_authentication, :connect
+
+  private
+    def connect
+      @client = GoogleCalendarService.new(current_user.oauth_token, current_user.refresh_token)
```
Comment (deepak) [inline app/controllers/api_controller.rb]: can use `||=` which will make this method idempotent

File: app/controllers/api_controller.rb [hunk 25]
```diff
@@ -0,0 +1,12 @@
...
+    end
+
+    def force_authentication
+      redirect_to login_path if session[:user_id].nil?
```
Comment (deepak) [inline app/controllers/api_controller.rb]: nothing wrong with `nil?` but `blank?` seems better in these cases
user is not found etc
not a strong preference. but makes our boundary check a little more robust

File: app/services/google_calendar_service.rb [hunk 26]
```diff
@@ -0,0 +1,126 @@
+class GoogleCalendarService
+  def initialize(oauth_token, refresh_token)
+    @client = Google::APIClient.new(:auto_refresh_token => true)
+    @client.authorization.access_token = oauth_token
+    @client.authorization.refresh_token = refresh_token
+    @client.authorization.client_id = ENV["GOOGLE_CLIENT_ID"]
+    @client.authorization.client_secret = ENV["GOOGLE_SECRET"]
+    if @client.authorization.refresh_token && @client.authorization.expired?
+      @client.authorization.fetch_access_token!
```
Comment (deepak) [inline app/services/google_calendar_service.rb]: can this method `fetch_access_token!` throw an error ?
if yes, it is a bad practice for `initialize` to throw an error
makes it harder to collect errors on "that" object

not a blocker for now, can work on it after you add tests for this class

File: app/services/google_calendar_service.rb [hunk 27]
```diff
@@ -0,0 +1,126 @@
...
+    response = @client.execute(api_method: service.events.list,
+                  parameters: { calendarId: "primary",
+                    timeMin: Time.now.beginning_of_month.iso8601,
+                    showDeleted: false,
+                    singleEvents: true,
+                    maxResults: 10,
+                    orderBy: "startTime"},
+                      headers: { "Content-Type": "application/json" })
+
+    items = response.data['items']
```
Comment (deepak) [inline app/services/google_calendar_service.rb]: this variable assignment is un-necesary

File: config/routes.rb (hunk 23 above)
Comment (HarshwardhanSingh) [inline config/routes.rb]: I added as: 'leave' because by default rails was creating /leaf instead of leaves.

File: app/services/google_calendar_service.rb (hunk 22 above)
Comment (HarshwardhanSingh) [inline app/services/google_calendar_service.rb]: didn't get you.

File: spec/controllers/leaves_controller_spec.rb (hunk 9 above)
Comment (HarshwardhanSingh) [inline spec/controllers/leaves_controller_spec.rb]: added some tests in https://github.com/code-mancers/leave-tracker/tree/writing_test_rspec

File: app/services/google_calendar_service.rb (hunk 26 above)
Comment (HarshwardhanSingh) [inline app/services/google_calendar_service.rb]: okay.

File: config/routes.rb (hunk 23 above)
Comment (deepak) [inline config/routes.rb]: yup. rails conventions :-)

Comment (HarshwardhanSingh) [inline config/routes.rb]: Should I rename it regardless that will name the route as **new_calendar_leafe_path** instead of **new_calendar_leave_path**.

Comment (deepak) [inline config/routes.rb]: ah man. misunderstood. thought you had made a typo in `leafe`
this is rails pluralization. did not expect this rule.

we can change it later.
leave it as-is for now
